server/log_rpc_server.go

```diff
@@ -587,3 +576,31 @@ func fetchNodesAndBuildProof(tx storage.LogTX, treeRevision, leafIndex int64, pr
 
 	return trillian.Proof{LeafIndex: leafIndex, ProofNode: proof}, nil
 }
+
+// getLeavesByHashInternal does the work of fetching leaves by either their raw data or merkle
+// tree hash depending on the supplied fetch function
+func (t *TrillianLogRPCServer) getLeavesByHashInternal(ctx context.Context, desc string, req *trillian.GetLeavesByHashRequest, fetchFunc func(storage.LogTX, [][]byte, bool) ([]trillian.LogLeaf, error)) (*trillian.GetLeavesByHashResponse, error) {
+	ctx = util.NewLogContext(ctx, req.LogId)
+	if len(req.LeafHash) == 0 || !validateLeafHashes(req.LeafHash) {
+		return &trillian.GetLeavesByHashResponse{Status: buildStatusWithDesc(trillian.TrillianApiStatusCode_ERROR, fmt.Sprintf("%s: Must supply at least one hash and none must be empty", desc))}, nil
+	}
+
+	tx, err := t.prepareStorageTx(req.LogId)
+
```

Comment:
Done. But I think it hides the important thing (assigning to tx) compared to the error handling.